drivers/ethernet/eth_smsc911x.c

```diff
@@ -0,0 +1,721 @@
+		return -1;
+	}
+
+	val |= 1 << 15;
+
+	if (smsc_phy_regwrite(SMSC9220_PHY_BCONTROL, val)) {
+		return -1;
+	}
+
+	return 0;
+}
+
+/**
+ * Advertise all speeds and pause capabilities
+ */
+void smsc_advertise_caps(void)
+{
+	u32_t aneg_adv = 0;
+
+	smsc_phy_regread(SMSC9220_PHY_ANEG_ADV, &aneg_adv);
+	aneg_adv |= 0xDE0;
+
+	smsc_phy_regwrite(SMSC9220_PHY_ANEG_ADV, aneg_adv);
+	smsc_phy_regread(SMSC9220_PHY_ANEG_ADV, &aneg_adv);
+}
+
+void smsc_establish_link(void)
+{
+	u32_t bcr = 0;
+	u32_t hw_cfg = 0;
+
+	smsc_phy_regread(SMSC9220_PHY_BCONTROL, &bcr);
+	bcr |= (1 << 12) | (1 << 9);
+	smsc_phy_regwrite(SMSC9220_PHY_BCONTROL, bcr);
+	smsc_phy_regread(SMSC9220_PHY_BCONTROL, &bcr);
+
+	hw_cfg = SMSC9220->HW_CFG;
+	hw_cfg &= 0xF0000;
+	hw_cfg |= (1 << 20);
+	SMSC9220->HW_CFG = hw_cfg;
+}
+
+inline void smsc_enable_xmit(void)
+{
+	SMSC9220->TX_CFG = 0x2 /*TX_CFG_TX_ON*/;
+}
+
+void smsc_enable_mac_xmit(void)
+{
+	u32_t mac_cr = 0;
+
+	smsc_mac_regread(SMSC9220_MAC_CR, &mac_cr);
+
+	mac_cr |= (1 << 3);     /* xmit enable */
+	mac_cr |= (1 << 28);    /* Heartbeat disable */
+
+	smsc_mac_regwrite(SMSC9220_MAC_CR, mac_cr);
+}
+
+void smsc_enable_mac_recv(void)
+{
+	u32_t mac_cr = 0;
+
+	smsc_mac_regread(SMSC9220_MAC_CR, &mac_cr);
+	mac_cr |= (1 << 2);     /* Recv enable */
+	smsc_mac_regwrite(SMSC9220_MAC_CR, mac_cr);
+}
+
+int smsc_init(void)
+{
+	unsigned int phyreset = 0;
+
+	if (smsc_check_id() < 0) {
+		return -1;
+	}
+
+	if (smsc_soft_reset() < 0) {
+		return -1;
+	}
+
+	smsc_set_txfifo(5);
+
+	/* Sets automatic flow control thresholds, and backpressure */
+	/* threshold to defaults specified. */
+	SMSC9220->AFC_CFG = 0x006E3740;
+
+#if 0
+	/* Not needed on QEMU, but may be needed on real hardware. */
+	if (smsc_wait_eeprom()) {
+		return -1;
+	}
+#endif
+
+	/* Configure GPIOs as LED outputs. */
+	SMSC9220->GPIO_CFG = 0x70070000;
+
+	smsc_init_irqs();
+
+	/* Configure MAC addresses here if needed. */
+
+	if (smsc_check_phy() < 0) {
+		return -1;
+	}
+
+	if (smsc_reset_phy() < 0) {
+		return -1;
+	}
+
+	k_sleep(PHY_RESET_TIMEOUT);
+	/* Checking whether phy reset completed successfully.*/
+	if (smsc_phy_regread(SMSC9220_PHY_BCONTROL, &phyreset)) {
+		return 1;
+	}
+
+	if (phyreset & (1 << 15)) {
+		return 1;
+	}
+
+	smsc_advertise_caps();
+	/* bit [12] of BCONTROL seems self-clearing. */
+	/* Although it's not so in the manual. */
+	smsc_establish_link();
+
+	/* Interrupt threshold */
+	SMSC9220->FIFO_INT = 0xFF000000;
+
+	smsc_enable_mac_xmit();
+	smsc_enable_xmit();
+	SMSC9220->RX_CFG = 0;
+	smsc_enable_mac_recv();
+
+	/* Rx status FIFO level irq threshold */
+	SMSC9220->FIFO_INT &= ~(0xFF);  /* Clear 2 bottom nibbles */
+
+	/* This sleep is compulsory otherwise txmit/receive will fail. */
+	k_sleep(2000);
+
+	return 0;
+}
+
+/* Driver functions */
+
+static enum ethernet_hw_caps eth_smsc911x_get_capabilities(struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T;
```

Comment:
Please consider to add VLAN support later. This is not preventing merging now but it would be good to have that supported.